snapshot/snapshotfs/upload.go

```diff
@@ -255,29 +255,52 @@ func (u *Uploader) uploadDir(ctx context.Context, rootDir fs.Directory, policyTr
 	return de, err
 }
 
-func (u *Uploader) foreachEntryUnlessCancelled(relativePath string, entries fs.Entries, cb func(entry fs.Entry, entryRelativePath string) error) error {
-	for _, entry := range entries {
-		if u.IsCancelled() {
-			return errCancelled
-		}
+func (u *Uploader) foreachEntryUnlessCancelled(ctx context.Context, parallel int, relativePath string, entries fs.Entries, cb func(ctx context.Context, index int, entry fs.Entry, entryRelativePath string) error) error {
+	type entryWithIndex struct {
+		entry fs.Entry
+		index int
+	}
 
-		entryRelativePath := relativePath + "/" + entry.Name()
+	ch := make(chan entryWithIndex)
+	eg, ctx := errgroup.WithContext(ctx)
 
-		if err := cb(entry, entryRelativePath); err != nil {
-			return err
+	// one goroutine to pump entries into channel until ctx is closed.
+	eg.Go(func() error {
+		defer close(ch)
+
+		for i, e := range entries {
+			select {
+			case ch <- entryWithIndex{e, i}: // sent to channel
+			case <-ctx.Done(): // context closed
+				return nil
+			}
 		}
-	}
+		return nil
+	})
 
-	return nil
-}
+	// launch N workers in parallel
+	for i := 0; i < parallel; i++ {
+		eg.Go(func() error {
+			for ewi := range ch {
+				if u.IsCancelled() {
```

Comment:
Given that this function may run for a long time, that is, process many entries, it seems that it should also check `ctx.Err()`

Separate PR: at some point it'd be good to use `context.Context` for upload cancellation as well.

Reply:
cancellation can't be driven off "ctx", because "ctx" is used to actually write to storage, and if we cancelled the parent context it would prevent any writes

Reply:
Understood. 
Thanks for the explanation.